### DISCUSSION TITLE
Has support for configuration of db-url for H2 been removed in release 21?

### DISCUSSION THREAD

**Discussion by barnemat-tieto** [keycloak/keycloak#20155] [Q&A]
I have been using Keycloak with H2, and specifying the db-url in the keycloak.conf file, for example:

`db-url=jdbc:h2:file:~/keycloak/data/h2/keycloakdb;AUTO_SERVER=TRUE`

With version 21 and 22, this no longer seems to work - see error messages below. If I omit the db-url property from the conf file, it works fine, and (after stopping keycloak) I can connect to and inspect the DB content using DBeaver with no problems.

Has support for configuration of db-url for H2 been removed with the upgrade to H2 v2.x in Keycloak 21?

Thanks.

```
2023-05-04 11:41:32,294 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode

2023-05-04 11:41:32,294 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to update database

#2023-05-04 11:41:32,294 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: liquibase.exception.MigrationFailedException: Migration failed for changeset META-INF/jpa-changelog-8.0.0.xml::8.0.0-updating-credential-data-not-oracle-fixed::keycloak:
     Reason: liquibase.exception.DatabaseException: Syntax error in SQL statement "UPDATE PUBLIC.CREDENTIAL SET CREDENTIAL_DATA = CONCAT('{""hashIterations"":', HASH_ITERATIONS, ',""algorithm"":""', ALGORITHM, '""}'), PRIORITY = '10', SECRET_DATA = CONCAT('{""value"":""', REPLACE([*]VALUE, '""', '\\\\""'), '"",""salt"":""__SALT__""}') WHERE TYPE = 'password' OR TYPE = 'password-history'"; expected "INTERSECTS (, NOT, EXISTS, UNIQUE, INTERSECTS"; SQL statement:
UPDATE PUBLIC.CREDENTIAL SET CREDENTIAL_DATA = CONCAT('{"hashIterations":', HASH_ITERATIONS, ',"algorithm":"', ALGORITHM, '"}'), PRIORITY = '10', SECRET_DATA = CONCAT('{"value":"', REPLACE(VALUE, '"', '\\"'), '","salt":"__SALT__"}') WHERE TYPE = 'password' OR TYPE = 'password-history' [42001-214] [Failed SQL: (42001) UPDATE PUBLIC.CREDENTIAL SET CREDENTIAL_DATA = CONCAT('{"hashIterations":', HASH_ITERATIONS, ',"algorithm":"', ALGORITHM, '"}'), PRIORITY = '10', SECRET_DATA = CONCAT('{"value":"', REPLACE(VALUE, '"', '\\"'), '","salt":"__SALT__"}') WHERE TYPE = 'password' OR TYPE = 'password-history']

```

**Comment by sschu** [comment]
Can you provide exact steps to reproduce?

**Reply by barnemat-tieto** [reply]
Sure.

1. Download and unzip keycloak, e.g. in D:\Testapps\keycloak\keycloak-21.1.1
2. Edit keycloak.conf (D:\Testapps\keycloak\keycloak-21.1.1\conf\keycloak.conf), adding the following properties:
```
db-url=jdbc:h2:D:/Testapps/keycloak/keycloak-21.1.1/data/h2/keycloakdb
db-username=sa
db-password=password
```
3. Run keycloak using kc start-dev 
`D:\Testapps\keycloak\keycloak-21.1.1\bin>kc start-dev`

**Reply by sschu** [reply]
I could indeed reproduce this on my windows system. Only seems to happen with file-backed h2. Afterwards I could also reproduce it under MacOS. So I guess it is not OS-related. I'll try to find out what's happening here.

**Reply by barnemat-tieto** [reply]
Thanks, much appreciated.

**Reply by barnemat-tieto** [reply]
Hi sschu,
Do you have any update on this problem? Is there a fix planned?
Thanks

**Reply by nabertrand** [reply]
@papandayan I was able to get this to work by adding `NON_KEYWORDS=VALUE` to the `db-url` parameter, e.g.
```
--db-url jdbc:h2:file:~/keycloak/data/h2/keycloakdb;AUTO_SERVER=TRUE;NON_KEYWORDS=VALUE
```
@sschu it looked to me like this option should be added automatically, but I'm not that familiar with the following code:
https://github.com/keycloak/keycloak/blob/904e688400cbb4470b15d88ef9f65f03ee1e7c71/quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java#L141-L157

**Reply by sschu** [reply]
Hmm, this code looks a bit strange since it adds a semicolon at the beginning and the end. I would assume it would be enough to add one in the beginning. This might be related to the problem, but not sure.

**Comment by samyvenkat** [comment]
This worked for me for the above error for version 25.06. Thanks.
